elastalert/ruletypes.py

```diff
@@ -720,6 +721,65 @@ def garbage_collect(self, ts):
                 placeholder.update({self.rules['query_key']: qk})
             self.handle_event(placeholder, 0, qk)
 
+class AdvQueryRule(RuleType):
+    """ A rule that uses a query_string query to perform a search """
+    required_options = frozenset(['alert_field'])
+
+    def __init__(self, *args):
+        super(AdvQueryRule, self).__init__(*args)
+        if 'max_threshold' not in self.rules and 'min_threshold' not in self.rules:
+            raise EAException("AdvQueryRule must have one of either max_threshold or min_threshold")
+        #self.query_string = self.rules.get('query_string')
+        self.rules['aggregation_query_element'] = {"query": ""}
+
+    def run_query(self):
+        # Implement the logic to run the query using the query_string
+        raise NotImplementedError()
+
+    def add_aggregation_data(self, payload):
+        for timestamp, payload_data in payload.items():
+            self.check_matches(payload_data,timestamp)
+    
+    def check_matches(self,data,timestamp):
+        #results=[]
+        for key, value in data.items():
+            if 'buckets' in value:
+                if len(value['buckets']) >=0 :
+                    self.check_matches_recursive(key,value['buckets'],timestamp)
+            else:
+                if self.crossed_thresholds(value['value']):
+                    match={"key":key,"value":value['value']}
+                    self.add_match(match)
+
+    def check_matches_recursive(self,data_key,data_value,timestamp,data_key_value=''):
+        key = data_key
+        if len(data_value) > 0:
+            for data in data_value:
+                local_data_key = data_key_value
+                data_key_value = data_key_value+ ','+ data['key'] if data_key_value else data['key']
+                for k,v in data.items():
+                    if k!= 'key' and k!= 'doc_count':
+                        if 'buckets' in v:
+                            local_key = key
+                            key = key + ','+k
+                            self.check_matches_recursive(key,v['buckets'],data_key_value)
```

Comment:
where are we passing the `timestamp` param here?

Reply:
Rectified